### PR TITLE
add use_global_stats option in batch norm layer

### PR DESCRIPTION
**Pull request by ascust**
Hi, I added the "use_global_stats" option in the batch norm layer to use the global mean and variance during the training instead of using the statistics within the batch. It is pretty useful when the batch is small and we want to finetune a pre-trained model.

### CODE REVIEW

File: src/operator/batch_norm-inl.h [hunk 1]
```diff
@@ -99,9 +102,18 @@ class BatchNormOp : public Operator {
       CHECK(req[batchnorm::kMean] == kNullOp || req[batchnorm::kMean] == kWriteTo);
       CHECK(req[batchnorm::kVar] == kNullOp || req[batchnorm::kVar] == kWriteTo);
       // The first three steps must be enforced.
-      mean = scale * sumall_except_dim<1>(data);
-      var = scale * sumall_except_dim<1>(F<mshadow_op::square>(
+      if (param_.use_global_stats)
+      {
```
Comment (tqchen) [inline src/operator/batch_norm-inl.h]: code style, ```if (param_.use_global_stats``

**Comment by tqchen** [comment]
The backward pass seems to be in-correct. Mainly because we do not need the batch-norm logic if it is a simple scale and add operator

**Comment by ascust** [comment]
@tqchen I followed the original paper. Like this:
screen shot 2016-03-21 at 7 15 59 pm
I think since the mean and var are constants, so the derivatives are supposed to be 0. I think caffe has the same implementation except caffe's version has only the batch norm part without scale and bias. What do you think?

**Comment by tqchen** [comment]
Because the mean and var are constants. The backprop to the bottom is even simpler. Since this actually becomes a broadcast multiply and broadcast add. You only need to rescale the gradient by the multiplication factor.

The original gradient is used in the case when the scaling factor is decided by the input(the variance)

**Comment by ascust** [comment]
7a9f24a6-ef99-11e5-998b-fc0b84eaea02
This is what I meant. I simply just set the derivatives to zero and did not change the code too much. The result would be multiplying by a scale just like you said. @tqchen

**Comment by tqchen** [comment]
@ascust  I created #2003 to support this. The main problem of current PR was due to not using the moving mean in place when during backprop with global stats, which is fixed in the proposed PR

**Comment by tqchen** [comment]
Thanks for proposing this change

**Comment by ascust** [comment]
@tqchen I have a question regarding this batchnorm layer. If I want to do data parallelisation and divide a batch into several mini batches. In this way, does each worker calculate the mean and var within this mini batch? If so, is there any way I can make the batchnorm layer calculate the mean and variance based on the whole batch?
